### DISCUSSION TITLE
Adding fonts by modifying system.img

### DISCUSSION THREAD

**Discussion by ShockedCoder** [chenxiaolong/avbroot#294] [Q&A] [answered]
I want to add [Fairfax](https://www.kreativekorp.com/software/fonts/fairfaxponahd) fonts to my phone.
I tried doing this through Termux and `sudo`, however that didn't work, and remounting it using `sudo mount -o remount,rw /system` didn't work as it didn't exist as a re-mountable drive.

Then I realized I had signed the OS with my own keys thanks to `avbroot`, I thought I could unpack the `system.img`, add the files in there, and the repack it.
However, I cannot repack it because of the size discrepancy between what `avb.toml` says and what `raw.img` is.

I'm a complete beginner at rooting and have no idea where to start searching for information regarding these sorts of things.

[OTA](https://release.calyxinstitute.org/FP5-ota_update-24507000.zip)

Here's the commands I executed and their outputs:

```sh
$ avbroot ota extract -i FP5-ota_update-24507000.zip -d extracted
  0.001s  INFO Extracting from the payload: boot, system, vbmeta, vbmeta_system, vendor_boot
  1.695s  INFO Successfully extracted OTA

$ cd extracted
$ avbroot avb unpack -i system.img
# [OUTPUT OMITTED] (will add, if needed)

$ resize2fs raw.img 1000M
resize2fs 1.47.0 (5-Feb-2023)
Resizing the filesystem on raw.img to 256000 (4k) blocks.
The filesystem on raw.img is now 256000 (4k) blocks long.

$ mkdir raw
$ doas mount raw.img raw
$ doas cp -a /usr/share/fonts/TTF/Fairfax* raw/system/fonts
$ doas umount raw
$ e2fsck -f raw.img # resize2fs requests this after every mount
e2fsck 1.47.0 (5-Feb-2023)
Pass 1: Checking inodes, blocks, and sizes
Pass 2: Checking directory structure
Pass 3: Checking directory connectivity
Pass 3A: Optimizing directories
Pass 4: Checking reference counts
Pass 5: Checking group summary information

/: ***** FILE SYSTEM WAS MODIFIED *****
/: 3072/3584 files (0.4% non-contiguous), 219385/256000 blocks

$ resize2fs -M raw.img # shrinks the fs to as small as it can 
resize2fs 1.47.0 (5-Feb-2023)
Resizing the filesystem on raw.img to 219426 (4k) blocks.
The filesystem on raw.img is now 219426 (4k) blocks long.

$ avbroot avb pack -o system.img
  1.285s  INFO Leaving AVB header unsigned
  4.030s ERROR Failed to write appended AVB image

Caused by:
    905834496 byte image size is too small to fit header or footer

```

**Comment by chenxiaolong** [comment, marked as answer]
I should add an option to have avbroot recompute the `image_size` field. In the meantime, you can make this work by changing `image_size` at the top of `avb.toml` to be ~32 MiB bigger than the size of your new `raw.img`.

(32 MiB is an arbitrary value that should work. Guessing a size that's too big is totally fine. It's only "wasting" space on the `super` partition, which isn't usable for anything else anyway.)

**Reply by chenxiaolong** [reply]
#296 introduces a new `--recompute-size` option to `avbroot avb pack`, which should make this easy. It'll be included in the next release of avbroot.

**Reply by ShockedCoder** [reply]
After adding the 32MiB, the packing worked, but it fails when I try to patch it.
```
  0.081s  INFO Replacing zip entry: META-INF/com/android/otacert
  0.081s  INFO Copying zip entry: apex_info.pb
  0.081s  INFO Copying zip entry: care_map.pb
  0.081s  INFO Patching zip entry: payload.bin
  0.081s  INFO Opening external image: system: "extracted/system.img"
  0.081s  INFO Extracting from original payload: vendor_boot
  0.156s  INFO Extracting from original payload: vbmeta_system
  0.156s  INFO Extracting from original payload: boot
  0.245s  INFO Extracting from original payload: vbmeta
  0.245s  INFO Patching boot images: boot, vendor_boot
  2.307s  INFO Patching system image: system
  3.285s  INFO Patched otacerts.zip offsets in system: [289591296..289592926]
  3.285s  INFO Patching vbmeta images: vbmeta_system, vbmeta
  3.339s ERROR Failed to patch OTA zip

Caused by:
    0: Failed to patch payload: payload.bin
    1: system descriptor (HashTree) in vbmeta_system must be a chain descriptor
```

Commands used:
```sh
#!/bin/dash

avbroot ota extract -i FP5-ota_update-24507000.zip -d extracted &&
cd extracted &&
avbroot avb unpack -i system.img &&
mkdir raw
resize2fs raw.img 1000M &&
doas mount raw.img raw &&
doas cp -a /usr/share/fonts/TTF/Fairfax* raw/system/fonts &&
doas umount raw &&
e2fsck -f raw.img &&
resize2fs -M raw.img &&
python -c 'import toml; avbtoml = toml.load("avb.toml"); avbtoml["image_size"] += 32*1024*1024; toml.dump(avbtoml, open("avb.toml", "w"))' &&
avbroot avb pack -o system.img -k ../avb.key -f --pass-env-var AVB_KEY_PASS &&
cd .. &&
avbroot ota patch -i FP5-ota_update-24507000.zip -o FP5-ota_update-24507000.patched.zip --replace system extracted/system.img --key-avb avb.key --key-ota ota.key --cert-ota ota.crt --magisk Magisk-v27.0.apk --magisk-preinit-device sda15 --pass-avb-env-var AVB_KEY_PASS --pass-ota-env-var OTA_KEY_PASS
```

EDIT: Signing the system.img seemed to be the issue

**Reply by chenxiaolong** [reply]
Glad you got it working!

> EDIT: Signing the system.img seemed to be the issue

Yep, I'd suggest removing `-f` and leaving `-k`. avbroot will sign if required or leave it unsigned if not.

**Reply by ShockedCoder** [reply]
Didn't quite get it working as intended, as I've found out that CalyxOS doesn't have a font switcher.

I heard replacing `Roboto-Regular.ttf` would do the trick, which it did, but it didn't look that nice since not many of my apps use the system font anyway, so I had to switch it back to normal 😆

At least I learned something along the way.
